src/lambda_toolshed/papillon.cljc

```diff
@@ -159,10 +155,21 @@
      by the chain)."
   ([ixs]
    (execute ixs {}))
-  ([ixs ctx]
+  #?(:org.babashka/nbb nil
+     :default
+     ([ixs ctx]
+      (let [c (chan 1)
+            handle  (fn [x] (go (>! c x)))
+            result (execute ixs ctx handle handle)]
+        (if (satisfies? Chrysalis result)
+          c
+          (do
+            (close! c)
+            result)))))
+  ([ixs ctx on-complete on-error]
    (let [ixs (map-indexed namer ixs)
          ctx (init-ctx ctx ixs)
          result (leave (enter ctx))]
-     (if (satisfies? ReadPort result)
-       (present-async result)
+     (if (satisfies? Chrysalis result)
```

Comment:
I don't think I like this...

Part of me says:

> If using callback functions, we should always present via callback functions, even if it ends up being all synchronous.

The other part:

> It is kinda nice to piggyback off the callback version, as that means we test that version when using the 2-arity version in our tests.  Also 🤢🤮 of callback functions when something is completely synchronous.

Reply:
Still trying to think through the best paths for all this. Like, is there a way to reasonably represent these concepts as part of the context map through reserved/predefined namespaced keys? `nbb` reader conditional can still happen, could be two separate inside of `present-async`.

Taking the path of "sane defaults, overwritable in user land" is a thing, always triggering callback functions if present does make sense. Don't want that functionality, don't set them.

Also, if they're represented as keys on the context, it allows for dynamically removing or adding them as needed during execution.

With the base problem being how best to represent an agnostic interface for async mechanisms, it has me wondering what path reduces the most conceptual complexity.

Reply:
The imposition of callback fns for completely synchronous work is a lot of incidental complexity that is not present in alternative interceptor libs.

Regarding the desire to support alternate implementations of deferred computations, has there been any investigation of an approach using `requiring-resolve` and multiple alternative impl namespaces (all implementing the Chrysalis protocol)  that could be optionally added as an additional artifact?  So require `lambda-toolshed/papillon` and `lambda-toolshed/papillon-manifold`.  In l-t/execute you could provide an optional parameter to override the default (core.async) and bring in an alternate implementation.

Apologies if this has already been considered...

Reply:
Interesting 🤔 

Pretty sure we chatted about something similar, but weren't really sure about how to represent it. one example @stevenproctor was looking at was (AsyncContext)[https://github.com/metosin/sieppari/blob/develop/src/sieppari/async.cljc#L8] from sieppari.

Reply:
Wow... that AsyncContext in sieppari is exactly what I had in mind, with the addition of separate artifacts to complement add-ons like manifold.

Reply:
I think the term is "Same Same, but Different." 😉 

I looked at it, and it was part of the reason I finally went with a distinct protocol.

I opted not to do `catch` and `continue` because of our error handling with `core.async` channels, and I went with only having `eclose` instead.

I put the core in `async`, although maybe if we decide to return it, we can move it to the main namespace. For extending ReadPort with the protocol, I put that in the `lambda-toolshed.papillon.async.core_async` namespace, which the tests require to load that functionality to get channels.

I left out `manifold` as I didn't want to put a dependency on `manifold` in the project, as those who want it could extend it themselves.

Given how we are currently handling the pure sync version, I was also unsure if we would want a blocking take similar to the `AsyncContext` of sieppari.

Reply:
My endorsement of sieppari's approach was limited to the protocol + extension to "built-in" types + (presumably) external artifacts/customization to extend to other defereds.  I did not mean to endorse `catch`/`continue` and TBH I have not dug deep enough into either sieppari's implementation or ours to render a useful opinion on the shape of the protocol.

(I will try to dig deeper into the Chrysalis implementation this weekend).

I strongly agree with leaving out any third party dependencies.  

I agree that a blocking take "behind the scenes" in the sync path is to be avoided.

Reply:
> The imposition of callback fns for completely synchronous work is a lot of incidental complexity that is not present in alternative interceptor libs.

Yeah, that is what I was thinking, but it did feel off that the result could be given via callback or return value, and the person would have to know which case they are in.

The other options that came to mind were:
1) don't piggyback the 2-arity and the 4-arity; if you use the 4-arity version, you always live in callback land. 😕 
1) to return the result and check if Chrysalis protocol, and have them call `eclose` with their handler (this was made cleaner by the `transition` from the `error-handling` branch, as before it took both the initial-context and the candidate context.
2) always return a Chrysalis, and make the raw context object (either the current `map` or a new record type) be shared so that `eclose` returns itself
3) return a Chrysalis as above, _and_ give a blocking take that does identify if it was in synchronous execution.
4) some other option that you all might think of

Reply:
IIUC, numbers 3/4 involve wrapping sync behavior in Chrysalis.  That's a reasonable solution as long as we stick to plain maps for context (with magic `type` metadata) not some custom type *and* we don't transition to async and then "fake it" with a blocking take -that will present an unexpected debugging nightmare.

Reply:
I pushed some changes that refined some of this thinking.

I now think of `present-async` as an abstraction, and we can `present-promise`, `present-channel`, or some other function if given in the context under the key `:lambda-toolshed.papillon/present-async`.

`present-promise` is only available in ClojureScript.

`present-channel` is not available in nbb.

The `present-async` default, if not given, is:
 - `present-promise` if in nbb,
 - `present-channel` otherwise.